fully connected(Dense Layer) operation:
1. 입력과 동일한 size의 필터(w)를 사용하여 하나의 출력 픽셀을 구성한다.
2. 각 출력 픽셀은 서로 다른 필터를 이용하여 생성한다.
3. 이로인해 필요한 필터의 수는 출력사이즈 * 입력사이즈 만큼 많은 필터를 요구한다.
4. 필터를 저장하기 위한 과도한 메모리를 요구한다.
5. 이미지의 독특한 성질(수평/수직/대각 픽셀간의 연속성)을 활용하지 못하는 문제가 발생

이미지의 독특한 성질:
1. 공간적인 연속된 정보를 가지고 있다.
2. 국부적으로 기억해야할 정보를 가지고 있다.
3. 채널간(BGR)에는 서로 독립된다.

convolutianal operation:
1. 입력의 사이즈와 무관한 필터(3x3, 5x5, 7x7, 1x3, 3x1, 1x1)를 사용한다.
2. 필터의 사이즈와 동등한 크기의 입력 이미지 영역으로부터 하나의 출력 픽셀을 구성한다.
3. 동일한 depth(channel) 출력 픽셀들(1개의 activation map)은 서로 필터를 공유한다. 
 - 즉, 하나의 필터가 재사용된다.
4. 필터를 저장하기 위한 메모리 소요량이 매우 적다.
5. 이미지의 공간적 특성을 잘 활용하고 있다.
6. 다채널 출력인 경우 채널간 필터를 공유하지 않는다.
7. 이미지의 가장자리 정보의 활용도를 높이기 위해 패딩 기술을 사용하다.

In [1]:
import numpy as np

In [2]:
img = np.arange(49, dtype=np.float).reshape((7,7))
img

<ipython-input-2-ad0a63674b4d>:1: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  img = np.arange(49, dtype=np.float).reshape((7,7))


array([[ 0.,  1.,  2.,  3.,  4.,  5.,  6.],
       [ 7.,  8.,  9., 10., 11., 12., 13.],
       [14., 15., 16., 17., 18., 19., 20.],
       [21., 22., 23., 24., 25., 26., 27.],
       [28., 29., 30., 31., 32., 33., 34.],
       [35., 36., 37., 38., 39., 40., 41.],
       [42., 43., 44., 45., 46., 47., 48.]])

In [11]:
fc_out = np.zeros_like(img)
fc_out

array([[0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.]])

In [12]:
cv_out = np.zeros_like(img)
cv_out

array([[0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.]])

In [4]:
# 기존 Dense 연산을 위해 필요한 필터의 갯수와 그 차원
fc_filter = np.random.randn(7,7,7,7)
print(fc_filter.shape)
fc_filter

(7, 7, 7, 7)


array([[[[ 3.82409871e-01, -8.45615199e-01,  5.50558267e-01, ...,
           1.06785883e+00, -8.07376831e-01, -1.47252378e+00],
         [-8.11349374e-01,  3.72784358e-01, -2.60426587e+00, ...,
           1.82553788e+00,  9.77147744e-01, -1.22251081e+00],
         [-3.98848867e-01,  2.99912246e-01,  3.21875335e-01, ...,
          -1.90228347e+00, -1.13407644e+00, -1.32013826e+00],
         ...,
         [ 1.56508489e+00,  7.32041100e-02, -3.33419199e-01, ...,
           1.28489382e+00,  6.41975605e-01,  4.33807331e-01],
         [ 1.02534281e-01, -9.61831760e-01,  2.00807048e+00, ...,
           6.64299092e-01, -5.61456213e-01,  5.92198026e-01],
         [-1.03766089e-01, -1.71273130e+00,  3.75379834e-01, ...,
          -7.08571749e-01, -2.29290812e+00,  1.30928695e+00]],

        [[ 7.70594104e-01, -1.28072658e-01, -1.49577894e+00, ...,
          -1.37864787e-02, -1.97488515e+00, -1.25198261e+00],
         [ 1.99965661e+00, -8.78584039e-02, -8.19097046e-01, ...,
          -2.01937129e

In [5]:
fc_bias = np.random.randn(7,7)
fc_bias

array([[-0.11558191, -1.13935728,  0.7408084 ,  0.00673201,  0.03981261,
        -0.87558218,  0.65271194],
       [-0.44045265, -0.76121959,  1.0265401 , -1.36097324, -0.77310083,
        -0.62805279,  1.52287536],
       [ 0.2433002 ,  0.87982486,  1.64795459, -1.66534343,  1.88094716,
         0.08869311, -0.42433417],
       [ 1.45101603,  1.42574357,  0.27004187,  0.61228867,  0.10743995,
        -0.42670809,  1.48173535],
       [-0.59696009,  0.1895014 ,  0.81976637,  0.20788036,  0.40313978,
         1.40304117, -0.40373195],
       [ 0.36857063, -1.99198992, -0.65666828, -0.16782116, -2.14545014,
         0.12485306,  0.79776321],
       [-0.3860442 ,  1.11736029, -1.88107436, -1.05509446,  0.18335082,
        -1.77648669,  1.44997705]])

In [6]:
7**4 + 7**2

2450

In [9]:
fc_filter[4, 6].shape

(7, 7)

In [10]:
for i in range(7):
    for j in range(7):
        fc_out[i,j] = np.sum(img * fc_filter[i,j]) + fc_bias[i, j]

NameError: ignored

In [8]:
fc_out

NameError: ignored

In [ ]:
fc_out.shape

(7, 7)

fc_out을 생성하기 위해 필요한 filter와 bias의 수는?
7\*7\*7\*7 + 7\*7 = 2401 + 49 = 2450

In [ ]:
img

array([[ 0.,  1.,  2.,  3.,  4.,  5.,  6.],
       [ 7.,  8.,  9., 10., 11., 12., 13.],
       [14., 15., 16., 17., 18., 19., 20.],
       [21., 22., 23., 24., 25., 26., 27.],
       [28., 29., 30., 31., 32., 33., 34.],
       [35., 36., 37., 38., 39., 40., 41.],
       [42., 43., 44., 45., 46., 47., 48.]])

$$\begin{align}
Oimg[i,j] &= \sum_{eliment}(Iimg[i:i+3, j:j+3] * F) + b \\
&= Iimg[i:i+3, j:j+3] \cdot F + b
\end{align}$$

padding이 없는 convolution 연산을 사용한다면:

In [ ]:
cv_filter = np.random.randn(3,3)
cv_filter

array([[ 2.36566373,  1.93692724,  0.01336884],
       [ 0.89723602,  0.37494582,  0.9497417 ],
       [ 0.20884162, -0.60924503, -0.00647824]])

In [ ]:
cv_bias = 0.1

In [ ]:
cv_nopadding_out = np.zeros((5,5))
cv_nopadding_out

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [ ]:
for i in range(5):
    for j in range(5):
        cv_nopadding_out[i, j] = np.sum(img[i:i+3, j:j+3]*cv_filter) + cv_bias

In [ ]:
cv_nopadding_out

array([[ 13.57301426,  19.70401595,  25.83501764,  31.96601933,
         38.09702102],
       [ 56.4900261 ,  62.62102779,  68.75202948,  74.88303118,
         81.01403287],
       [ 99.40703794, 105.53803964, 111.66904133, 117.80004302,
        123.93104471],
       [142.32404979, 148.45505148, 154.58605317, 160.71705486,
        166.84805656],
       [185.24106163, 191.37206332, 197.50306502, 203.63406671,
        209.7650684 ]])

cv_out에 필요한 filter와 bias의 수는?
- 3\*3 + 1 = 10

In [ ]:
cv_padding_out = np.zeros((7,7))
cv_padding_out

array([[0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.]])

In [ ]:
conv_img = np.zeros((9,9))
conv_img

array([[0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0.]])

In [ ]:
conv_img[1:-1, 1:-1] = img.copy()
conv_img

array([[ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  1.,  2.,  3.,  4.,  5.,  6.,  0.],
       [ 0.,  7.,  8.,  9., 10., 11., 12., 13.,  0.],
       [ 0., 14., 15., 16., 17., 18., 19., 20.,  0.],
       [ 0., 21., 22., 23., 24., 25., 26., 27.,  0.],
       [ 0., 28., 29., 30., 31., 32., 33., 34.,  0.],
       [ 0., 35., 36., 37., 38., 39., 40., 41.,  0.],
       [ 0., 42., 43., 44., 45., 46., 47., 48.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.]])

In [ ]:
for i in range(7):
    for j in range(7):
        cv_padding_out[i, j] = np.sum(conv_img[i:i+3, j:j+3]*cv_filter) + cv_bias
cv_padding_out

array([[ -3.26679945,  -1.09594387,   0.71909802,   2.5341399 ,
          4.34918179,   6.16422367,   1.42176907],
       [  1.70931916,  13.57301426,  19.70401595,  25.83501764,
         31.96601933,  38.09702102,  30.97410012],
       [ 20.32414147,  56.4900261 ,  62.62102779,  68.75202948,
         74.88303118,  81.01403287,  67.19468586],
       [ 38.93896379,  99.40703794, 105.53803964, 111.66904133,
        117.80004302, 123.93104471, 103.4152716 ],
       [ 57.55378611, 142.32404979, 148.45505148, 154.58605317,
        160.71705486, 166.84805656, 139.63585733],
       [ 76.16860843, 185.24106163, 191.37206332, 197.50306502,
        203.63406671, 209.7650684 , 175.85644307],
       [124.96034932, 248.71747604, 255.25535938, 261.79324273,
        268.33112608, 274.86900942, 234.30805797]])

### blur
인접픽셀과의 평균값으로 출력시킨다.

In [ ]:
img = np.arange(49).reshape((7,7))
img

array([[ 0,  1,  2,  3,  4,  5,  6],
       [ 7,  8,  9, 10, 11, 12, 13],
       [14, 15, 16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25, 26, 27],
       [28, 29, 30, 31, 32, 33, 34],
       [35, 36, 37, 38, 39, 40, 41],
       [42, 43, 44, 45, 46, 47, 48]])

In [ ]:
fsize = 3
filter = np.full((fsize, fsize), 1/(fsize*fsize))
filter

array([[0.11111111, 0.11111111, 0.11111111],
       [0.11111111, 0.11111111, 0.11111111],
       [0.11111111, 0.11111111, 0.11111111]])

In [ ]:
out = np.zeros((5,5))
bias = 0
for i in range(5):
    for j in range(5):
        out[i, j] = np.sum(img[i:i+fsize, j:j+fsize]*filter) + bias
        
out

array([[ 8.,  9., 10., 11., 12.],
       [15., 16., 17., 18., 19.],
       [22., 23., 24., 25., 26.],
       [29., 30., 31., 32., 33.],
       [36., 37., 38., 39., 40.]])

In [ ]:
import numpy as np
X = np.random.randint(0, 256, (3,6,6))
X

array([[[219, 239, 206, 154, 187,  72],
        [133,  90, 230, 239,  37, 151],
        [ 21, 196,  37,  75,  85, 191],
        [169, 211, 153, 241,  97, 172],
        [141, 109, 202,  96,  98, 184],
        [125, 208, 205, 217, 118, 242]],

       [[102, 212, 232, 104, 101, 130],
        [ 89, 222, 180, 169,  99, 245],
        [188,  20, 250,  70, 207, 227],
        [247, 174,  38,  31, 220, 159],
        [  2, 184, 120, 145,  54, 118],
        [122, 112, 174, 234, 142,  71]],

       [[ 50,  70, 232,  23, 140, 103],
        [ 73, 105,  11, 143, 184,  17],
        [253, 217, 248, 115,  39,  60],
        [139,  47, 246, 250, 219, 117],
        [168, 126, 109,  19, 143,  53],
        [163, 208, 222, 194, 135,  56]]])

In [ ]:
output = np.zeros((3,3,3), np.int)
output

<ipython-input-5-d5c8b8e90dcb>:1: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  output = np.zeros((3,3,3), np.int)


array([[[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]]])

In [ ]:
i = 0; j=1; k=2
X[i, 2*j:2*(j+1), 2*k:2*(k+1)]

array([[ 85, 191],
       [ 97, 172]])

In [ ]:
output.shape[1]/2

1.5

In [ ]:
for i in range(output.shape[0]):
    for j in range(output.shape[1]):
        for k in range(output.shape[-1]):
            # print(i, j)
            output[i,j,k] = np.max(X[i, 2*j:2*(j+1), 2*k:2*(k+1)])

output

array([[[239, 239, 187],
        [211, 241, 191],
        [208, 217, 242]],

       [[222, 232, 245],
        [247, 250, 227],
        [184, 234, 142]],

       [[105, 232, 184],
        [253, 250, 219],
        [208, 222, 143]]])

In [ ]:
X

array([[[219, 239, 206, 154, 187,  72],
        [133,  90, 230, 239,  37, 151],
        [ 21, 196,  37,  75,  85, 191],
        [169, 211, 153, 241,  97, 172],
        [141, 109, 202,  96,  98, 184],
        [125, 208, 205, 217, 118, 242]],

       [[102, 212, 232, 104, 101, 130],
        [ 89, 222, 180, 169,  99, 245],
        [188,  20, 250,  70, 207, 227],
        [247, 174,  38,  31, 220, 159],
        [  2, 184, 120, 145,  54, 118],
        [122, 112, 174, 234, 142,  71]],

       [[ 50,  70, 232,  23, 140, 103],
        [ 73, 105,  11, 143, 184,  17],
        [253, 217, 248, 115,  39,  60],
        [139,  47, 246, 250, 219, 117],
        [168, 126, 109,  19, 143,  53],
        [163, 208, 222, 194, 135,  56]]])